### PR TITLE
Add monitoring configuration from provider specific components

### LINKED ISSUE #1351: Monitoring and logging configuration for extension controllers
**What would you like to be added**:
While progressing with #308 we have moved out most of the provider-specific information of the Gardener code. However, there are still left-overs for monitoring and logging configuration of provider-specific components, e.g. for the cloud-controller-manager ([metric scrape config](https://github.com/gardener/gardener/blob/ff1c8d1a39528aee41d5ed7867f6b2b05011a8b2/charts/seed-monitoring/charts/core/charts/prometheus/templates/config.yaml#L318-L345), [alerting rules](https://github.com/gardener/gardener/blob/2d3f929213150ab88d25de684eb4573f0c7107c9/charts/seed-monitoring/charts/core/charts/prometheus/rules/cloud-controller-manager.rules.yaml), [dashboards](https://github.com/gardener/gardener/blob/bb4b16a385ec3d7c59df3ebcacfa6e704e793e29/charts/seed-monitoring/charts/grafana/dashboards/owners/kubernetes-control-plane-status-dashboard.json#L323), and [logging parse configuration](https://github.com/gardener/gardener/blob/2ea4dc7d7cb1797f96373fcf6073dda49ebbad6c/charts/seed-bootstrap/charts/fluentd-es/templates/fluent-bit-configmap.yaml#L85-L90)).
We should also move these parts to the provider-specific extension code, because only the extensions know these components and how they are scraped/displayed/... in the best way.

**Why is this needed**:
Part of #308, see motivation in [GEP-1](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md).

**Implementation proposal**:
The general idea is that the extension controllers, as part of the control loops that deploy provider-specific components, may write a `ConfigMap` (one or many) into the shoot namespace in the seed. This `ConfigMap` contains the respective information for Gardener's logging and monitoring stacks. It must be labelled in a special way such that Gardener finds them.
During the steps in Gardener's control loops that are deploying the logging/monitoring configuration it finds all such `ConfigMap`s, extends the default/non-provider-specific configuration, and finally deploys the logging/monitoring stacks.

#### Logging

In case of logging the parse information is always the same and not shoot-specific. Also, the configuration is part of `fluent-bit` which is deployed once per seed and scrapes all containers by default.
Here, the mentioned `ConfigMap` should be created in the `garden` namespace in the seed (instead of in any shoot namespace).
Gardener's [seed controller deploys the logging stack](https://github.com/gardener/gardener/blob/master/pkg/operation/seed/seed.go#L267-L470) regularly (in every sync period, default `30s`).

In the of the cloud-controller-manager, every extension controller deploys a `ConfigMap` into the `garden` namespace that contains the filter/parse configuration for CCM:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: extension-controlplane-logging-ccm
  namespace: garden
  labels:
    extensions.gardener.cloud/configuration: logging
data:
  parse_config: |
    [FILTER]
        Name                parser
        Match               kubernetes.cloud-controller-manager*cloud-controller-manager*
        Key_Name            log
        Parser              kubeapiserverParser
        Reserve_Data        True
```

Before deploying `fluent-bit` Gardener reads all such configmaps and extends the default config.

#### Monitoring

In case of monitoring the scrape/alerting/dashboard information is sometimes shoot-specific. 
Here, the mentioned `ConfigMap` should be created in the `shoot--<project>--<name>` namespace in the seed.
Gardener's [shoot controller deploys the logging stack](https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/controlplane.go#L260-L411) at the end of every shoot reconciliation flow.

In the of the cloud-controller-manager, every extension controller deploys a `ConfigMap` into the `shoot--<project>--<name>` namespace that contains the scrape/alerting/dashboard configuration for CCM:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: extension-controlplane-monitoring-ccm
  namespace: shoot--project--name
  labels:
    extensions.gardener.cloud/configuration: monitoring
data:
  scrape_config: |
    scrape_configs:
    - job_name: cloud-controller-manager
      scheme: https
      tls_config:
        insecure_skip_verify: true
        cert_file: /etc/prometheus/seed/prometheus.crt
        key_file: /etc/prometheus/seed/prometheus.key
      honor_labels: false
      kubernetes_sd_configs:
      - role: endpoints
        namespaces:
          names: [shoot--project--name]
      relabel_configs:
      - source_labels:
        - __meta_kubernetes_service_name
        - __meta_kubernetes_endpoint_port_name
        action: keep
        regex: cloud-controller-manager;metrics
      # common metrics
      - action: labelmap
        regex: __meta_kubernetes_service_label_(.+)
      - source_labels: [ __meta_kubernetes_pod_name ]
        target_label: pod
      metric_relabel_configs:
      - process_max_fds
      - process_open_fds

  alerting_rules:
    groups:
    - name: cloud-controller-manager.rules
      rules:
      - alert: CloudControllerManagerDown
        expr: absent(up{job="cloud-controller-manager"} == 1)
        for: 15m
        labels:
          service: cloud-controller-manager
          severity: critical
          type: seed
          visibility: all
        annotations:
          description: All infrastruture specific operations cannot be completed (e.g. creating loadbalancers or persistent volumes).
          summary: Cloud controller manager is down.

  dashboard_operators:
    <some-json-describing-a-grafana-dashboard-for-operators>

  dashboard_users:
    <some-json-describing-a-grafana-dashboard-for-users>
```

Before deploying `prometheus`, `alertmanager`, and `grafana` Gardener reads all such configmaps and extends the default config.

##### Details

* The labels on the `ConfigMap`s should be exposed as constants in Gardener's `core` API group.
* All the data keys in the `ConfigMap`s should be optional.
* @dkistner and @wyb1 suggested to create new/dedicated dashboards for the provider-specific components and not to try patching/merging their information into the default dashboards.

**Definition of done**:
- [x] All the [provider extensions](https://github.com/gardener/gardener-extensions/tree/master/controllers) deploy such `ConfigMap`s as part of their `controlplane` reconcile loops
- [ ] Gardener picks up all such `ConfigMap`s before deploying the logging/monitoring stack and extends the default configuration.
- [ ] The contract for logging/monitoring has been documented for developers in the `docs/extensions` folder.

### PR DESCRIPTION
**Pull request by svetlinas**
**What this PR does / why we need it**:
Every extension controller has deployed many `ConfigMap`s into the shoot namespace in the seed cluster. These `ConfigMap`s contain the respective information for Gardener's monitoring stacks.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/gardener/issues/1351

**Special notes for your reviewer**:
Depends on https://github.com/gardener/gardener-extensions/pull/344

**Release note**:

```improvement operator
Gardener does now read `ConfigMap`s labelled with `extensions.gardener.cloud/configuration=monitoring` and injects their data into the `prometheus` and `grafana` configurations. This allows extension controllers to define their provider-specific monitoring configuration for the components they deploy.
```

### CODE REVIEW

File: pkg/operation/botanist/monitoring.go [hunk 1]
```diff
@@ -36,13 +38,33 @@ import (
...
+		prometheusHost      = b.ComputePrometheusHost()
+		alertingRules       = strings.Builder{}
+		scrapeConfigs       = strings.Builder{}
+		operatorsDashboards = strings.Builder{}
+		usersDashboards     = strings.Builder{}
 	)
 
+	// Find extensions provider-specific monitoring configuration
+	existingConfigMaps := &corev1.ConfigMapList{}
+	if err := b.K8sSeedClient.Client().List(context.TODO(), existingConfigMaps,
```
Comment (rfranzke) [inline pkg/operation/botanist/monitoring.go]:
```suggestion
	if err := b.K8sSeedClient.Client().List(ctx, existingConfigMaps,
```

**Review by rfranzke — COMMENTED** [review]
Very nice, thanks for the PR, just a minor thing.

Over to @wyb1 and @dkistner for checking and evaluating.

**Review by wyb1 — COMMENTED** [review]
Since you are removing the alert tests, we should move them to the extensions. This can be done in a later PR though.

**Comment by svetlinas** [comment]
> Since you are removing the alert tests, we should move them to the extensions. This can be done in a later PR though.

Should we open an issue so that this could not be forgotten?

**Comment by wyb1** [comment]
@svetlinas I will open an issue. 
/lgtm

**Review by rfranzke — APPROVED** [review]

**Pull request merged**